module/zfs/spa.c

```diff
@@ -7430,6 +7430,89 @@ spa_vdev_add(spa_t *spa, nvlist_t *nvroot, boolean_t check_ashift)
+ * 6. 2nd spare is kicked in for new blank disk
+ * 7. At this point two spares are kicked in for the original disk1.
+ *
+ * It looks like this:
+ *
+ * NAME                                            STATE     READ WRITE CKSUM
+ * tank2                                           DEGRADED     0     0     0
+ *   draid2:6d:10c:2s-0                            DEGRADED     0     0     0
+ *     scsi-0QEMU_QEMU_HARDDISK_d1                 ONLINE       0     0     0
+ *     scsi-0QEMU_QEMU_HARDDISK_d2                 ONLINE       0     0     0
+ *     scsi-0QEMU_QEMU_HARDDISK_d3                 ONLINE       0     0     0
+ *     scsi-0QEMU_QEMU_HARDDISK_d4                 ONLINE       0     0     0
+ *     scsi-0QEMU_QEMU_HARDDISK_d5                 ONLINE       0     0     0
+ *     scsi-0QEMU_QEMU_HARDDISK_d6                 ONLINE       0     0     0
+ *     scsi-0QEMU_QEMU_HARDDISK_d7                 ONLINE       0     0     0
+ *     scsi-0QEMU_QEMU_HARDDISK_d8                 ONLINE       0     0     0
+ *     scsi-0QEMU_QEMU_HARDDISK_d9                 ONLINE       0     0     0
+ *     spare-9                                     DEGRADED     0     0     0
+ *       replacing-0                               DEGRADED     0    93     0
+ *         scsi-0QEMU_QEMU_HARDDISK_d10-part1/old  UNAVAIL      0     0     0
+ *         spare-1                                 DEGRADED     0     0     0
+ *           scsi-0QEMU_QEMU_HARDDISK_d10          REMOVED      0     0     0
+ *           draid2-0-0                            ONLINE       0     0     0
+ *       draid2-0-1                                ONLINE       0     0     0
+ * spares
+ *   draid2-0-0                                    INUSE     currently in use
+ *   draid2-0-1                                    INUSE     currently in use
+ *
+ * ARGS:
+ *
+ * newvd:  New spare disk
+ * pvd:    Parent vdev_t the spare should attach to
+ *
+ * This function returns B_TRUE if adding the new vdev would create a double
+ * spare condition, B_FALSE otherwise.
+ */
+static boolean_t
+spa_vdev_new_spare_would_cause_double_spares(vdev_t *newvd, vdev_t *pvd)
+{
+	vdev_t *ppvd;
+	const char *ppvd_op, *pvd_op;
+
+	pvd_op = pvd->vdev_ops->vdev_op_type;
+
+	ppvd = pvd->vdev_parent;
+	if (ppvd == NULL || ppvd->vdev_ops == NULL)
+		return (B_FALSE);
+
+	ppvd_op = ppvd->vdev_ops->vdev_op_type;
+	if (ppvd_op == NULL)
+		return (B_FALSE);
+
+	/*
+	 * To determine if this configuration would cause a double spare, we
+	 * look at the vdev_op_type string of the parent vdev, and of the
+	 * parent's parent vdev.  We also look at vdev_isspare on the new disk.
+	 * A double spare condition looks like this:
+	 *
+	 * 1. parent of parent's vdev_op_type is "spare" or "dspare"
+	 * 2. parent's vdev_op_type is  "replacing"
+	 * 3. new disk is a spare
+	 */
+	if ((strcmp(ppvd_op, VDEV_TYPE_SPARE) == 0) ||
+	    (strcmp(ppvd_op, VDEV_TYPE_DRAID_SPARE) == 0))
+		if (strcmp(pvd_op, VDEV_TYPE_REPLACING) == 0)
```

Comment:
The code below prefers to do it as `pvd->vdev_ops == &vdev_replacing_ops` etc.  Though this does not look wrong and I don't have specific preference outside uniformity.

Reply:
Ok, I'll switch to that in my next push.